server/docs/design/bidi-producer-consumers-streaming.md

```diff
@@ -0,0 +1,339 @@
+# Bi-directional Producer/Consumer Streaming with gRPC
+
+## Purpose
+
+A primary use case of the `hedera-block-node` is to stream live BlockItems (see Terms section) from a producer 
+(e.g. Consensus Node) to N consumers (e.g. Mirror Node) with the lowest possible latency while correctly preserving the 
+order of the BlockItems.  This document outlines several possible strategies to implement this use case and the design 
+of the recommended approach. All strategies rely on the Helidon 4.x.x server implementations of HTTP/2 and gRPC 
+services to ingest BlockItem data from a producer and then to stream the same BlockItems to downstream consumers.  It
+does this by defining bidirectional gRPC streaming services based on protobuf definitions.
+
+Helidon provides well-defined APIs and extension points to implement business logic for these services. The main entry 
+point for custom logic is an implementation of `GrpcService`.
+
+---
+
+## Goals
+
+1) Consumers must be able to dynamically subscribe and unsubscribe from the live stream of BlockItems emitted by the
+    producer.
```

Comment:
Interesting -- if I am a consumer and I subscribe, do I get whatever the next block item is, or do I start from the beginning of a block?

Reply:
This implementation was for the "live stream" so the consumer would get the newest BlockItems in-flight.  It sounds like there are thousands of BlockItems per block.  Is the correct behavior to always start the stream by emitting the first BlockItems from the current Block and then catch up to the "live stream"?  There's a future story where the consumer can specify the block number they'd like to start from and they'd receive continuous stream until they catch up to the "live stream".

Reply:
I was expecting we would not start them till the first item of next block. Otherwise sending useless data they are paying for.

Reply:
If a new consumer subscribes and just misses the start of a Block, what's the anticipated latency until they'd start seeing BlockItems from the next Block? Is there a danger they think the stream "isn't working" because they're not seeing data immediately flowing by and they retry connections in frustration?

Reply:
Latency should be a second or less if they miss substantially the entire block.
Current blocks are about one per second, and may be more frequent with future gossip improvements.

Reply:
@jsync-swirlds Ok, makes sense.  

So, @jasperpotts and @rbair23, if I understand the choices, we have:

1) The system will wait until the next Block to start live streaming BlockItems - consumers will accept the latency

2) The system will immediately stream BlockItems on the current Block to backfill missed items until it catches up to the live stream of BlockItems - consumers will accept some stale BlockItems until the system catches up.

Which one should we go with?

Reply:
Given feedback in the demo, I've updated the doc to support option #1